patroni/validator.py

```diff
@@ -758,6 +758,52 @@ def assert_(condition: bool, message: str = "Wrong value") -> None:
     assert condition, message
 
 
+class EnumValidator(object):
+    """Validate a setting against a list of allowed values
+
+    ivar: allowed: a list of allowed values
+    ivar: case_sensitive: set to True to do case sensitive comparisons
+    """
+
+    def __init__(self, allowed: List[Any], case_sensitive: bool = False):
+        """Create an :class:`EnumValidator` object with the given rules.
+
+        :param allowed: list of allowed values
+        :param case_sensitive: set to True to do case sensitive comparisons
+
+        >>> EnumValidator(allowed=[])(3)
+        False
+        >>> EnumValidator(allowed=[2,3,5])(3)
+        True
+        >>> EnumValidator(allowed=[2,3,'something'])(4)
+        False
+        >>> EnumValidator(allowed=['off','oN','blind'])('ON')
+        True
+        >>> EnumValidator(allowed=['off','on','blind'], case_sensitive=True)('ON')
+        False
+        >>> EnumValidator(allowed=[True])('on')
+        True
+        >>> EnumValidator(allowed=[True])('off')
+        False
+        >>> EnumValidator(allowed=[True,False])('1')
+        True
+        """
+        self.case_sensitive = case_sensitive
+
+        if self.case_sensitive:
+            self.allowed = allowed
+        else:
+            self.allowed = [a.lower() if isinstance(a, str) else a for a in allowed]
+
+    def __call__(self, value: Any) -> bool:
+        """Check if *value* is part of the allowed values."""
+
+        if not self.case_sensitive and isinstance(value, str):
+            value = value.lower()
+
+        return value in self.allowed or parse_bool(value) in self.allowed
```

Comment:
Sorry, but I really dislike that the `EnumValidator` for some reason also handling boolean values.
Enums in Postgres are generally strings, with some minor exceptions like `synchronous_commit`, where PG treats `1` the same way as `on` or `tRuE`. 

- For `wal_level` we don't expect anything but only string values. Neither we do expect to push more GUCS via command line in a foreseeable future. A very simple `EnumValidator` with `patroni.collections.CaseInsensitiveSet()` will perfectly do the job of validating `wal_level` values.
- For `track_commit_timestamp` and `wal_log_hints` we expect only booleans (even if they arrive as `str` or `int` objects).

Mixing two different things together in a single class is not a good practice.
For the reference, there is already [patroni/postgresql/validator.py](https://github.com/zalando/patroni/blob/d46ca88e6bb4b4788ec86689026c2d2e4da68c79/patroni/postgresql/validator.py#L114-L136), that makes a difference between GUC's that are just enums and GUC's that are a mixture of enum with bool, but unfortunately it can't be reused.